File: pharmacy-core-backend/app/schemas/billing.py
"""Pydantic schemas for the Billing domain — the HTTP contract.

Input/output split, same convention as the Medicine domain:
- BillingRequest  — what the CLIENT sends (just the free-text sentence)
- BillingResponse — what the SERVER returns (the receipt: sale id, lines, total, errors)

The graph's internal state (BillingState) is NOT exposed directly — these schemas
are the clean public face. Internal fields (batch_id, medicine_id) are included
because they're useful to the pharmacist UI, but cost_price etc. never appear.
"""
from pydantic import BaseModel, Field, field_validator

from app.core.phone import InvalidPhoneNumberError, normalize_indian_mobile

# Shared bounds for days supply. 1..365 matches the CHECK constraint on both
# sale_items.days_supply and medicines.default_days_supply -- Pydantic gives the
# friendly 422, the database is the backstop that also catches direct SQL.
DAYS_SUPPLY_MIN = 1
DAYS_SUPPLY_MAX = 365


class BillingRequest(BaseModel):
    """Input contract: the pharmacist's free-text order.

    FastAPI validates incoming JSON against this. Empty / oversized input → 422
    automatically, before the graph (and the paid LLM call) ever runs.
    """

    pharmacist_input: str = Field(
        ...,
        min_length=1,
        max_length=1000,
        description="Free-text order, e.g. '2 strips Crocin 500mg for Anurag 9876543210'",
    )


class MatchCandidate(BaseModel):
    """A possible catalog medicine for an uncertain (voice-mis-heard) line."""

    medicine_id: int
    name: str


class BillingLineItem(BaseModel):
    """One line on the returned receipt — mirrors a priced_item from graph state."""

    name: str
    quantity: int
    unit: str
    medicine_id: int
    batch_id: int
    batch_number: str
    expiry_date: str
    unit_price: float
    line_total: float

    # M6.1 -- the medicine's typical course length, so the billing UI can
    # pre-fill the days-supply box. A HINT for the form only: whatever the
    # pharmacist finally confirms is what gets written to the sale line.
    # None means the catalogue has no typical duration for this medicine, which
    # is the honest answer for anything taken as-needed.
    default_days_supply: int | None = None

    # ----- Voice-match metadata (defaults keep older callers working) -----
    # True when the match was uncertain and the owner should confirm before billing.
    needs_confirm: bool = False
    # The raw spoken text we matched FROM (e.g. "mat CB"), for "did you hear this right?".
    matched_from: str | None = None
    # Alternative catalog medicines the owner can switch to (dropdown).
    candidates: list[MatchCandidate] = Field(default_factory=list)


class BillingResponse(BaseModel):
    """Output contract: the receipt the server returns.

    sale_id is None when no sale could be created (e.g. every medicine was
    out of stock or unrecognized). In that case `errors` explains why and the
    router returns HTTP 422.
    """

    sale_id: int | None = Field(
        default=None,
        description="The new invoice id, or null if no sale could be created",
    )
    total_amount: float = Field(default=0.0, description="Grand total of all line items")
    customer_name: str | None = Field(default=None)
    customer_phone: str | None = Field(default=None)
    items: list[BillingLineItem] = Field(
        default_factory=list,
        description="One entry per successfully-billed medicine line",
    )
    errors: list[str] = Field(
        default_factory=list,
        description="Soft errors / warnings (e.g. 'Medicine not found', 'Insufficient stock')",
    )


class ConfirmLineItem(BaseModel):
    """One owner-reviewed line sent to POST /confirm to finalize the sale.

    Note there is NO price field here: the server RE-FETCHES the MRP from the DB
    at confirm time and recomputes the price. The client can never dictate the
    price (server-side pricing rule). These fields are exactly the 'batched_item'
    shape the quote step returned, minus the prices.
    """

    name: str
    quantity: int = Field(..., ge=1, le=1000)
    unit: str
    medicine_id: int
    batch_id: int
    batch_number: str
    expiry_date: str

    # M6.1 -- how many days this line is expected to last.
    #
    # Optional, and omitting it is a real answer, not a lazy one: it records
    # "the pharmacist does not know", and the future refill engine must skip
    # that line rather than invent a duration. `ge=1` means a client cannot send
    # 0 to mean unknown -- zero days of supply is not a thing, and allowing it
    # would give us two spellings of unknown that behave differently downstream.
    days_supply: int | None = Field(
        default=None, ge=DAYS_SUPPLY_MIN, le=DAYS_SUPPLY_MAX
    )


class PriceItemRequest(BaseModel):
    """Price ONE medicine by id (used when the owner picks a different candidate
    from a confirm dropdown). Server fetches the FEFO batch + MRP."""

    medicine_id: int
    quantity: int = Field(..., ge=1, le=1000)
    name: str | None = Field(default=None)
    unit: str | None = Field(default="strip")


class ConfirmSaleRequest(BaseModel):
    """Input contract for POST /confirm — finalize a previously-quoted bill.

    The owner may have edited quantities or removed rows in the preview; we
    persist exactly the items sent here. Customer fields stay optional.
    """

    items: list[ConfirmLineItem] = Field(..., min_length=1)
    customer_name: str | None = Field(default=None)
    customer_phone: str | None = Field(default=None)

    # M6.1 -- explicit WhatsApp opt-in, captured at the counter.
    #
    # Defaults to False, and that default is the whole point: consent is never
    # implied by a customer handing over a phone number. Under the DPDP Act the
    # number is given for the sale, and reusing it to message them is a separate
    # purpose needing its own affirmative act.
    #
    # Only meaningful alongside a phone number; ignored otherwise, since there is
    # no customer row to attach it to.
    whatsapp_opt_in: bool = Field(default=False)

    @field_validator("customer_phone")
    @classmethod
    def _normalize_phone(cls, value: str | None) -> str | None:
        """Canonicalise the phone, or reject it -- see app/core/phone.py.

        Done here so every caller of /confirm gets the same treatment and the
        UNIQUE index on customers.phone sees one spelling per human. A bad number
        fails with 422 rather than being stored: the field is optional, so the
        pharmacist can simply clear it and save the sale.
        """
        try:
            return normalize_indian_mobile(value)
        except InvalidPhoneNumberError as e:
            raise ValueError(str(e)) from e

File: pharmacy-frontend/src/lib/api/billing.js
/**
 * Billing API.
 *
 * The bill is built one line at a time. For each line the SERVER picks the batch
 * (FEFO) and sets the price from the database — the browser never sends a price,
 * and `confirm` recomputes everything again before writing the sale. A tampered
 * client payload cannot change what a customer is charged.
 *
 * There is no LLM in this flow.
 */
import { getJSON, postJSON } from "./client";

/** Units a line can be sold in. The backend defaults to "strip" when unset. */
export const UNIT_OPTIONS = ["strip", "tablet", "bottle", "tube", "pack", "unit"];

/** The medicine catalog the picker searches. */
export function listCatalog() {
    return getJSON("/api/v1/medicines");
}

/**
 * Price one line: medicine + quantity -> FEFO batch, unit price, line total.
 *
 * 422 means the medicine has no usable batch — unexpired and in stock. That is a
 * normal outcome worth showing the user, not an error to swallow.
 */
export function priceLine({ medicineId, quantity, name, unit }) {
    return postJSON("/api/v1/billing/price-item", {
        medicine_id: medicineId,
        quantity,
        name,
        unit,
    });
}

/**
 * Finalize the reviewed bill. Writes the sale and decrements stock. 201 on success.
 *
 * M6.1 adds two facts the future refill engine needs, and nothing that acts on
 * them: how long each line is expected to last, and whether the customer agreed
 * to WhatsApp reminders. Nothing here sends a message.
 */
export function confirmSale(items, customerName, customerPhone, whatsappOptIn = false) {
    return postJSON("/api/v1/billing/confirm", {
        items: items.map((it) => ({
            name: it.name,
            quantity: it.quantity,
            unit: it.unit,
            medicine_id: it.medicine_id,
            batch_id: it.batch_id,
            batch_number: it.batch_number,
            expiry_date: it.expiry_date,
            // null means "the pharmacist does not know how long this lasts", and
            // the backend stores that NULL rather than guessing. An empty string
            // or 0 would both be wrong: 0 fails validation, "" is not an int.
            days_supply: toDaysSupply(it.days_supply),
        })),
        customer_name: customerName || null,
        customer_phone: customerPhone || null,
        // Consent is never implied by the customer giving a phone number.
        whatsapp_opt_in: Boolean(whatsappOptIn),
    });
}

/**
 * Coerce the days-supply input to what the API expects.
 *
 * An <input type="number"> yields a STRING, and an emptied one yields "". Both
 * have to become either a real integer or an explicit null, because the backend
 * distinguishes them: null is recorded as "unknown, do not remind", while a
 * number schedules a future reminder. Sending "" would be a 422; sending 0 is
 * rejected by design so that unknown has exactly one spelling.
 */
export function toDaysSupply(value) {
    if (value === null || value === undefined || value === "") return null;
    const parsed = Number(value);
    if (!Number.isInteger(parsed) || parsed < 1) return null;
    return parsed;
}
